contracts/Genesis.sol

```diff
@@ -82,107 +91,107 @@ contract Genesis is ERC721, VRFConsumerBase, Ownable {
-            addressToMintCount[msg.sender] < MAX_PER_MINT,
-            "No more minting spot left"
-        );
-
+        require(tokenMintedCount[msg.sender] == 0, "Already minted");
+        tokenMintedCount[msg.sender]++;
         _nextTokenId.increment();
-        bytes32 requestId = requestRandomNumberForTokenId(mintIndex);
-        addressToMintCount[msg.sender] = addressToMintCount[msg.sender].add(1);
+        requestId = requestRandomNumberForTokenId(mintIndex);
         emit RequestedRandomNFT(requestId);
         return requestId;
     }
 
-    function requestRandomNumberForTokenId(uint256 _tokenId)
-        internal
-        returns (bytes32 requestId)
-    {
-        require(
-            LINK.balanceOf(address(this)) >= fee,
-            "Not enough LINK - fill contract with faucet"
-        );
+    /**
+     * Requests a random number to Chainlink VRF for a given token id
+     * @param tokenId id of the token
+     * @return requestId id of the request sent to Chainlink
+     */
+    function requestRandomNumberForTokenId(uint256 tokenId) internal returns (bytes32 requestId) {
+        require(LINK.balanceOf(address(this)) >= fee, "Not enough LINK");
         requestId = requestRandomness(keyHash, fee);
         requestIdToSender[requestId] = msg.sender;
-        requestIdToTokenId[requestId] = _tokenId;
+        requestIdToTokenId[requestId] = tokenId;
         return requestId;
     }
 
     /**
-     * Randomization
+     * Generate a leaf of the Merkle tree with a nonce and the address of the sender
+     * @param nonce nonce to be used
+     * @param addr id of the token
+     * @return leaf generated
      */
-    function generateRandomTraits(uint256 _randomNumber)
-        private
-        returns (TokenType genesisType)
-    {
-        require(
-            godsCount + demiGodsCount + elementalsCount > 0,
-            "All NFTs have been generated"
-        );
+    function generateLeaf(uint256 nonce, address addr) internal pure returns (bytes32 leaf) {
+        return keccak256(abi.encodePacked(nonce, addr));
+    }
+
+    /**
+     * Verifies the proof of the sender to confirm they are in the whitelist
+     * @param nonce nonce to be used
+     * @param proof proof
+     * @return valid TRUE if the proof is valid, FALSE otherwise
+     */
+    function verifyProof(uint256 nonce, bytes32[] memory proof) internal view returns (bool valid) {
+        return MerkleProof.verify(proof, merkleTreeRoot, generateLeaf(nonce, msg.sender));
+    }
+
+    /**
+     * Returns the token type (god, demi-god, or elemental) given a random number
+     * @param randomNumber random number provided
+     * @return tokenType a randomly picked token type
+     */
+    function getTokenType(uint256 randomNumber) private returns (TokenType tokenType) {
+        require(godsCount + demiGodsCount + elementalsCount > 0, "All NFTs have been generated");
         uint256 totalCountLeft = godsCount + demiGodsCount + elementalsCount;
```

Comment:
The reason I had done this the other way was to "lower" fees on an error. Does that make sense?
I guess the whole function was also using more gas because of it?

Reply:
I didn't touch this function

Reply:
Well, check the code, it was definitely changed :)
Minor anywya